server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java

```diff
@@ -560,6 +560,11 @@ public void removeSegment(DataSegment segment, DataSegmentChangeCallback callbac
             },
             this::resolveWaitingFutures
         );
+      } else if (status.get().getState() == Status.STATE.SUCCESS) {
```

Comment:
I feel a better approach could be to augment the possible states to denote that the last success was a Drop success rather than invalidating here. (e.g. Maybe we could add a new state SUCCESS_DROPPED, etc.)

The current approach could cause a problem if we send three subsequent requests to Load a segment.
- Request 1 would honor the request and mark it as PENDING. The request would then complete and be marked SUCCESS.
- Request 2 would invalidate the status of the previous one.
- Request 3 would again try to Load the Segment, even though it's already loaded.

Please correct me if my understanding is not right.

Reply:
Segment loads are idempotent, the node can be asked to load the same segment many times, it will only load it once.  If it gets asked multiple times, it will generate some log spew, but won't create a correctness bug.  The cache here is just an optimization on top of that idempotency to keep things out of synchronized blocks when not needed.  The intent of the change is to "invalidate" the cache as soon as the node believes that it has accurately reported back a meaningful success response.

Reply:
Note that this is not just an issue with dropping segments, but an issue in the load case too.

Reply:
@kfaraz  what is the harm if the segment is attempted to be loaded again? maybe wasteful, if it occurs, but how often do you think this would happen?

Reply:
If we only invalidate the cache after the second call to reload then that call itself is a no-op right? However, since there was already a drop before it, that call should not have been a no-op and should reload right?. In this case we would have to do something like a load-drop-load-load to actually reload the segment?

Reply:
So, the thing that is going on here is that the HTTP protocol allows for batches of change requests.  The request will return as soon as any one of those requests completes, in a batch of 5, 1 might complete immediately while others are still running.  When 1 completes, the coordinator will issue a new request.  In the meantime, one of the ones that was requested might have completed, this will cache that it completed so that when the new request comes it can immediately respond back saying that it completed loading, this causes another request for load from the coordinator to happen.

So, even in the normal case, once `SUCCESS` (meaning that the request has been completed) has been achieved, the need for having the thing in the cache is no more.  This change is making those semantics explicit.

There are still potential corner cases (as there are with any distributed protocol).  Those corner cases are covered by the idempotency of the API.  If, due to some corner case, the same thing gets request multiple times, then it will only be loaded or dropped once (once it has been loaded, there's nothing to load.  Once it has been dropped, there's nothing more to drop).

The coordinator assignment algorithm also doesn't expect that a request to load/drop will actually succeed.  It starts over from fresh state every time it wakes up because failures can happen.  This means that if something is lost, it will identify the correct current state and eventually recover on its own.  I.e. even if there is some point in time when a "load-drop-load-load" occurs, that's not actually a problem because at the end it will actually loaded.  There is only a problem if it goes into an infinite loop of "load-load-load" without ever actually making progress on the work.

Reply:
@cheddar , for the coordinator assignment algorithm that you noted, im confused how it protects against a segment never being loaded properly in the face of the bug that this pr is fixing. The coordinator will likely notice that a particular segment that was asked to be loaded hasnt been loaded, so it may request that the historical load the segment again, but if the historical still has the cache entry for an earlies load request for the same segment, how would it ever load the segment?

Reply:
Thanks for clarifying @cheddar . Makes sense